### ISSUE TITLE
Message data model discussion

### ISSUE THREAD

**Issue by folletto** [WordPress/wp-feature-notifications#16]
We need to collaborate across design and dev to clarify what's the data model we want to use. Having the data model helps inform design, in return design will check if the data model works for all our use cases.

The current draft is in code [here](https://github.com/WordPress/wp-notify/blob/develop/includes/Notification.php), translated in simpler terms, this is the i1 draft:

* `get_id` int
* `timestamp` timedate
* `sender` user_id
* `recipients` users? roles? mix?
* `message` text? html?

Let's review what we need to do, how much it matches the [requirements document](https://docs.google.com/document/d/1SoIaFqXkXiVzq5mizbZQafMfL36bD0WKj8iwYM823MI/edit), and how well it fits the use cases explored in the designs.

---

### Jump to Current Update
The discussion evolved, but no need to read through everything, just check the latest summary and proposal here:
https://github.com/WordPress/wp-notify/issues/16#issuecomment-724608463

**Comment by folletto** [comment]
Reporting here a note from the [i1 design iteration](https://github.com/WordPress/wp-notify/issues/1#issuecomment-690501831): that design iteration has a lot going on in terms of data model, with optional fields an all. It might not be what we want, but here's the current needs to make that design possible:

* All messages have a `message` content
* Some `message` content is HTML
* Optional `title`
* Optional `action` (label + URL?)
* Optional `image` (direct SVG or URL)

Open for discussion.

**Comment by folletto** [comment]
Reporting here also @daniloercoli's comment

—

Just my 2 cents about the data model from a mobile native perspective.

* If we go with HTML content in the body, i hope it will be restricted to few tags, otherwise it could be hard to represent the body properly in a native web view.
* Action - What the button should do on mobile native? i guess the easiest approach will be to open a web view, both in app or on system browser, and take the user to the proper HTML page within wp-admin.

**Comment by raaaahman** [comment]
>     * All messages have a `message` content
> 
>     * Some `message` content is HTML
> 
>     * Optional `title`
> 
>     * Optional `action` (label + URL?)
> 
>     * Optional `image` (direct SVG or URL)

I like the idea to put image and action in the data model, rather than putting it in the HTML message. This could really ease the communication between frontend and backend IMO.

Although, I would say we should focus on simple messages without any options for v0.1, in order to get a proof of concept ready the sooner we can.

**Comment by raaaahman** [comment]
>     * If we go with HTML content in the body, i hope it will be restricted to few tags, otherwise it could be hard to represent the body properly in a native web view.

That's a good point. Do you know about any resource we may use to list the tags we can use, and those we shouldn't?

**Comment by folletto** [comment]
All good points. 

It seems to me however that implementing HTML might end up being way more costly (a large set of elements to support and/or to filter) than a couple of extra optional fields in the payload, at least for the first version. 
Would this be a correct assumption?

**Comment by raaaahman** [comment]
I'd say this is a legitimate concern. 

Although, to have compatibility with the existing notices that have been set by plugins and themes, HTML support will be mandatory.

Whether we focus on this task early or late is to be discussed for sure.

**Comment by jonathanbossenger** [comment]
Some thoughts.

I don't think that WP Notify has to be compatible with existing notices set by plugins and themes. What we're building here will eventually become the defacto way to manage notifications, and so plugin and theme authors should be required to create notifications on whatever way we decide they should work, for the best user experience for the user. Therefore, what we should consider building this so that it's easier for a plugin or theme author to create a new notification with the same functionality that is currently possible with admin_notices, but using WP Notify's functionality.

So, in my opinion, the message field should NOT support HTML in the first iteration. What we can do is allow for an `action_link` field, so that a plugin or theme developer can add a URL to the field, which allows the user to trigger any internal or external action required. When a plugin or theme registers a new notification, the `action_link` field should only accept a valid URL, either generated using admin_url() for internal actions, or hard coded to some external action.

I agree with having a title field, but making it optional.

I've started a database-schema.md document, which you can view here, and we can update as we agree on the schema. https://github.com/WordPress/wp-notify/blob/feature/database-schema/docs/database-schema.md

**Comment by jonathanbossenger** [comment]
I'm not opposed to adding the image field, as an optional field, as well. wp-notify should be opinionated about where that image is displayed, not the plugin/theme author.

**Comment by jonathanbossenger** [comment]
Additional question, as it pertains to the sender and recipients fields

Should this be an email/list of emails, or a user id/list of user ids? Do we want/need to allow notifications to be sent to users not registered in the system?

**Comment by raaaahman** [comment]
> I don't think that WP Notify has to be compatible with existing notices set by plugins and themes. What we're building here will eventually become the defacto way to manage notifications, and so plugin and theme authors should be required to create notifications on whatever way we decide they should work, for the best user experience for the user. Therefore, what we should consider building this so that it's easier for a plugin or theme author to create a new notification with the same functionality that is currently possible with admin_notices, but using WP Notify's functionality.

In my opinion, there is two possible way of bringing WP Notify into the Core:
 
1. Either Wp Notify is compatible with the old admin notices and then replaces them
2. Either both systems coexists and WP Notify will slowly become the standard way to go because we provide more features / ease of use. 

You seem to suggest we go with option 2, am I right?

> So, in my opinion, the message field should NOT support HTML in the first iteration. 

In the first iteration, definitely not. In the end though, it will only be possible if we choose option 2 (see above).

> I agree with having a title field, but making it optional.
> I'm not opposed to adding the image field, as an optional field, as well. wp-notify should be opinionated about where that image is displayed, not the plugin/theme author.

I agree here, standard notification with optional fields sounds less pain to manage. Even if this would probably mean adding more features over the lifecycle of the component.

**Comment by raaaahman** [comment]
> I've started a database-schema.md document, which you can view here, and we can update as we agree on the schema. https://github.com/WordPress/wp-notify/blob/feature/database-schema/docs/database-schema.md

I'll quote myself from October 12, 2020 Office Hours (pretentious I know :D ):

> As far as I understand it, @folletto, which is part of the designers in the team, is probably only concerned by the data to display to the user
> Mapping this data directly to a database table scheme seems to be going a step too far into implementation details as:
> - Not all notifications have a need to be persistent
> - We'll probably start off with non persistent notifcations, as these are the existing notifications we want to replace
> I believe that between the database and the view, the data should be filtered between a few abstraction layers, for two reasons:
> 1. Not all data are relevant to be displayed: the recipient will be well aware that she IS the recipient of this notification, for example
> 2. We would probably want to let website developers swap the database storage for something else in the future, like a Redis cache or something else, at their very own discretion.

**Comment by jonathanbossenger** [comment]
> You seem to suggest we go with option 2, am I right?

That's kind of the only way we can do it. There are too many existing plugins and themes that do things the old way, and to force them to switch once the new way is introduced may be tricky. The way I see it is, once we have defined the new way, and it's merged into core, at that point the plugin and theme teams can require any new plugins or themes to use the new system, and give existing plugin developers some time to migrate over. But for a time at least I see there being a state where both are still possible.

**Comment by folletto** [comment]
Trying to recap what I got from the exchange above, would this be correct?

* Required: email/list of emails, or a user id/list of user ids [TBD]
* Required: message body (plain text, not HTML)
* Optional: title
* Optional: action (label + URL)
* Optional: image (direct SVG or URL)

It also needs a way to track which plugin or component is making the request. How would that work?

I'd also want to highlight back from the requirements that we need to cover both on-page messages and notification messages — I'm just stating this again as I'm not 100% sure what "non persistent" implies, to be sure.

**Comment by folletto** [comment]
Quick i3 update made during the Slack chat:

* Required: list of references to recipients [TBD separately]
* Required: message body (plain text, not HTML)
* Optional: title
* Optional: action (label + URL)
* Optional: image (direct SVG or URL)

**Comment by raaaahman** [comment]
> * Optional: action (label + URL)

I'd rather have two separate entities than a composed one. I propose:

* Optional: action label (plain text, single line)
* Optional: action link (URL)

> * Optional: image (direct SVG or URL)

We'd probably better keep the data types strict, an URL (local or distant) may as well be used for SVG, PNG or JPG, so I'd favor taking only URL.

> It also needs a way to track which plugin or component is making the request. How would that work?

Probably just a sender data:

* Optional: sender (string identifier)

I envision a string identifier akin to those for text domains. Optional because not technically required for the notification system to operate, but highly recommended.

> I'd also want to highlight back from the requirements that we need to cover both on-page messages and notification messages — I'm just stating this again as I'm not 100% sure what "non persistent" implies, to be sure.

I would say that both notifications should be very similar in the data they carry. The difference should be made mostly in the way they are triggered and displayed. The persistence is another consideration (notification may be registered fora later read, typically the "Would you rate my plugin?" notification that stays from plugin activation until the user publishes a rating -or dismisses them-).

**Comment by folletto** [comment]
> I'd rather have two separate entities than a composed one.

In my head that was a `{ image: "", url: "" }` kind of thing, just for the sake of explanation, I agree shouldn't be a single field. :) 

> We'd probably better keep the data types strict, an URL (local or distant) may as well be used for SVG, PNG or JPG, so I'd favor taking only URL.

I'd suggest to look into Guetenberg handling. I suggested the above as it's what it's doing, if I recall correctly, and it adds some flexibility (with relative ease of detection). But I'm ok either way to be clear, I just wanted to state the source of the advice.

> Optional: sender (string identifier)
> Optional because not technically required for the notification system to operate, but highly recommended.

This is important, I'm not sure how to handle this... the problem here is that with something like this we lose the ability to control the notifications. In short, we can't build a notification panel where we can configure notifications by Plugin source if we don't have something like this. Also: spoofing — what if one plugin "spoofs" a notification from another plugin?

So in short, we need something to:

1. Allow notifications configuration / control panel / settings
2. Avoid spoofing

How can we do that?

> I would say that both notifications should be very similar in the data they carry. The difference should be made mostly in the way they are triggered and displayed. 

Yes and no... if you can see from the designs, _Notification_ types can be just a text message and that's it, while _ON page_ types are likely to have the full load. Maybe however we are saying the same thing, in the sense that we agree on what's optional, so yes, either way the optionals have to be handled in both scenarios.

**Comment by abuyoyo** [comment]
> I don't think that WP Notify has to be compatible with existing notices set by plugins and themes. 

That's very disconcerting. And it's a significant BC-break.
It's also an _avoidable_ BC-break. Adding PHP output buffering to existing/deprecated `admin_notice` hook(s) can collect all existing notices' HTML for use within the new feature.

**Comment by jonathanbossenger** [comment]
> Quick i3 update made during the Slack chat:
> 
>     * Required: list of references to recipients [TBD separately]
> 
>     * Required: message body (plain text, not HTML)
> 
>     * Optional: title
> 
>     * Optional: action (label + URL)
> 
>     * Optional: image (direct SVG or URL)

In the spirit of keeping this moving along, may I suggest we start with this as the agreed upon data structure, and iterate as needed

- recipients (required, comma separated list of user ids)
- message (required, plain text, no HTML allowed)
- title (text, optional)
- action (url, optional)
- action_label (text, optional)

Things like images, or moving the recipients list to (possibly) it's own table later on can be done as we see the need for it.

**Comment by jonathanbossenger** [comment]
> > I don't think that WP Notify has to be compatible with existing notices set by plugins and themes.
> 
> That's very disconcerting. And it's a significant BC-break.
> It's also an _avoidable_ BC-break. Adding PHP output buffering to existing/deprecated `admin_notice` hook(s) can collect all existing notices' HTML for use within the new feature.

@abuyoyo I think there may be some misunderstanding of the goal of WP Notify.

The way I see it, there will be 3 phases of implementing this into core

1. WP Notify is added to core, the plugin and theme developers are given the ability to register WP Notify notifications, which display in a central WP Notify notifications hub. 
2. Once there is a defined platform for registering non admin_notice hooked notification, the theme and plugin review teams can create guidelines requiring theme and plugin developers to use the new system, and replace all their usage of admin_notices for notifications. Any "on page" notifications (ie those that are specifically shown only on a specific page, and contextual to the page content) can remain using admin_notices.
3. Once theme and plugin developers are comfortable with the new system, and the only notifications that are registered using admin_notices are "on page" notifications, we can then work towards version 2, where we update how theme and plugin developers use admin_notices for "on page" notifications.

**Comment by folletto** [comment]
> Things like images, or moving the recipients list to (possibly) its own table later on can be done as we see the need for it.

There's [already a need](https://github.com/WordPress/wp-notify/issues/1#issuecomment-690501831) for images tho. ;)

**Comment by jonathanbossenger** [comment]
Aha, apologies, I missed that. For the sake of expediency then:

 - recipients (required, comma separated list of user ids)
 - message (required, plain text, no HTML allowed)
 - title (text, optional)
 - action (url, optional)
 - action_label (text, optional)
 - image (url, optional, should be part of the WP install (ie theme or plugin))

**Comment by folletto** [comment]
Ok, we're aligned — that sounds good to me right now.

**Comment by abuyoyo** [comment]
@jonathanbossenger 
What you have described above is a shift in project goals for feature-notifications/wp-notify.

1. The main impetus behind a WP feature-notifications was to _solve_ current on-page echo-your-own notification system. As well as provide a modern UI/modal for the display of these notifications. Your comments here suggest this is not the case and that feature-notifications/wp-notify is in fact a _break_ from current (and past) admin notice system. It's just _another_ interface for notices.
2. Providing multiple concurrent notification displays - some on-page, and some in a modal is a bad user interface and a bad user experience. If feature-notifications/wp-notify does not address the current printing of messages (excluding in-line notices - those should stay obviously) then it's not solving what it set out to solve.
3. Your last comment completely sidesteps the issue of backward-compatibility of old plugins. Unless I misread this comment of yours, and "version 2" is a phase _before_ integration into core. In which case my last comment is moot.
> we can then work towards version 2, where we update how theme and plugin developers use admin_notices for "on page" notifications.

**Comment by folletto** [comment]
Note: I think it's an important discussion, and might be worth having a separate ticket to explore backward compatibility and rollout, so it doesn't get mixed up with this thread about data model. :)

**Comment by raaaahman** [comment]
>     * recipients (required, comma separated list of user ids)
> 
>     * message (required, plain text, no HTML allowed)
> 
>     * title (text, optional)
> 
>     * action (url, optional)
> 
>     * action_label (text, optional)
> 
>     * image (url, optional, should be part of the WP install (ie theme or plugin))

Sorry, this still bothers me:

- The `recipients` isn't a data to be displayed directly to the end user: whether she sees the notifications, or she doesn't. She should be oblivious to the fact that it is a list of WP Users, which is a strictly technical concern. And defining now that it should be a list of WP Users would create obstacles for us to change that later: if for example we want to send direct emails to users. My opinion is to keep it abstract from the notification model: a WP Notify Recipient id, where the WP Notify Recipient is another model separated from the WP Notify Notification model
- The same goes with images, the fact that it is a URL neither serves the design team, that will create mock ups with raster / vector images, nor the development team, that might want to switch for a base64 encoded string or embedded svg later on, if performance become an issue. At this point, a WP Notify Image id would be sufficient information for both teams. Then our WP Notify Image code may use the better suited approach to display it.

This might look like a lot of nitpicking at the moment (and in a way, it is), but abstracting things upfront to allow us more flexibility is a philosophy I wanted to explore in this project.

**Comment by raaaahman** [comment]
@abuyoyo Thank you for sharing your concerns. Indeed, there might be a long delay before we will be able to totally replace the old notification system.

Hi-jacking current admin notices to redirect them to the notification center might be doable, although, we'd have to focus our efforts. If we have to take time to fix compatibility issues with the many ways the admin notices have been used, this is time that will not go to tailor the main WP Notify experience.

Once the user experience is great with WP Notify, there will probably be users wanting to make the switch completely. This will indeed be a good time to consider the total replacement of old admin notices.

**Comment by schlessera** [comment]
Some background on what my thinking was while I worked on the code in here:
- the data model for the base notification is "enforced", so it should only contain the data that ALL notifications will share without fail. Additional data per notification should be possible, but it should work via an extension mechanism that can be added to via plugins and it should be able to be "scaled down" again to the base notification so that all notifications still can be displayed in all contexts.
- the design was specifically not meant to be a drop-in replacement for the existing admin notices, as these notices are per their very fundamental design flawed and problematic. There should of course be a migration path for plugins and themes to move from old admin notices to the new notification system, but backward compatibility should not be considered, as it will from the start taint the new design.

**Comment by jonathanbossenger** [comment]
> @jonathanbossenger
> What you have described above is a shift in project goals for feature-notifications/wp-notify.

@abuyoyo I don't necessarily agree with that statement. The main project goal for feature-notifications/wp-notify has always been:

`Create a new (better) way to manage and deliver notifications to the relevant audience.`

Backward compatibility was not something that was a project goal. 

I would also refer to our meeting earlier this year, where we discussed the possiblity of working with the current admin_notices implementation (https://wordpress.slack.com/archives/C2K1C71FE/p1578927343063500) and it was agreed that any work done to attempt to salvage admin_notices would not be useful. 

As @schlessera suggests out in that meeting "Work on a solid, scalable and manageable solution for notification channels. Then moving what people use admin_notices for right now to these notification channels will solve the current problem as a side-effect."

So while I agree our impetus was to solve current on-page echo-your-own notification system, it was not to solve this by working within the current admin_notices framework, which I think we can all agree is not suited to the task.

**Comment by jonathanbossenger** [comment]
> This might look like a lot of nitpicking at the moment (and in a way, it is), but abstracting things upfront to allow us more flexibility is a philosophy I wanted to explore in this project.

@raaaahman I'd rather you nitpick now, than we only realise this later, when it's too late. :-)

One question I would raise at this point, is the effect of adding multiple new tables to the WordPress database. If I'm reading your comments correctly, you're effectively suggesting a `notifications` table, and a `notification_recipients` table. Under normal circumstances (ie managing our own project) this would probably not be a problem, but I have no idea how the rest of the core team feels about adding multiple tables to the database?

That being said, should we rather just move forward with our specific requirements, and cross that bridge when we get there?

I'm trying to make sure we balance building the best notification systems we can with making it more straight forward to merge it into core later. But maybe I'm over thinking things?

**Comment by raaaahman** [comment]
> One question I would raise at this point, is the effect of adding multiple new tables to the WordPress database. If I'm reading your comments correctly, you're effectively suggesting a `notifications` table, and a `notification_recipients` table. Under normal circumstances (ie managing our own project) this would probably not be a problem, but I have no idea how the rest of the core team feels about adding multiple tables to the database?

Model =/= relational database tables. There is many ways to store data in an application, starting with in memory storage (no persistence). If we think in tables and SQL queries that early, we may end up coupling our design to much with a technology that might be problematic to switch later on. Which I hope to avoid.

> That being said, should we rather just move forward with our specific requirements, and cross that bridge when we get there?

That's what I'm thinking at the moment. The brainstorm as moved us far enough, the details might get clearer when we'll get our hands dirty.

> I'm trying to make sure we balance building the best notification systems we can with making it more straight forward to merge it into core later. But maybe I'm over thinking things?

I would say that is a reasonable philosophy, although we have to expect WordPress to change until we get to merge things.

**Comment by folletto** [comment]
In the design discussion, @hedgefield raised a good point here https://github.com/WordPress/wp-notify/issues/26#issuecomment-741740545: what if we specify one data field just for the icon?
This woulda allow to either pass the icon (for both on-hub and on-page) but also add an extra image (which allows on-page to have both icon and image).

Thoughts?

**Comment by raaaahman** [comment]
What do you want the icon to represent? The sender (WordPress Core or Plugin / Theme)? In that case the icon should be part of the Sender model.

**Comment by folletto** [comment]
There are different approaches to that. One is that one: the icon is set and yes it's the sender. The on-page will always show it, and the on-hub can override it if an image is set.

**Comment by schlessera** [comment]
As per [meeting in Slack](https://wordpress.slack.com/archives/C2K1C71FE/p1617720265066100):

This is a "less is more" situation. You can easily extend the set of required fields later on if you can provide a fallback value for the consumers that have not implemented that field. But you can never remove any fields once they were published.

Also, the extension mechanism could be built in such a way that if older code tries to extend with a field addition that was later added as a default field, it would not break but just use the default field instead. This way, you can monitor what the usual extensions are, and if an extension seems to be always done because it happens to be crucial, you can "promote" that to a built-in field in a later version.

This explicitly takes into account future compatibility just as well as backward compatibility.

This is similar to making everything `final` and `private` as a first default and only opening up later on if there is a pressing need. Once something is part of the public interface, you cannot easily undo that.

Note that we can just as well provide multiple "implementations" apart from the `WPNotify_BaseNotification` from the get-go if that makes sense. So, maybe a notification with an image is a separate implementation `WPNotify_IllustratedNotification` or so...? This implementation would then also use the field extension mechanism to add the needed fields to make the image happen.

**Comment by Luehrsen** [comment]
I might want to add a few thought about some requirements I face in my daily life considering notifications. I draw my inspiration mainly from notification-systems like Xenforo and Facebook, but also push-notification systems.

1) Notification Levels: Different notifications have different importance. I'd like a mechanism that reflects that. We could draw inspiration from the myriad of debug systems out there and define levels like `Error`, `Security`, `Info`, etc. 

The intent here is to make sure that an important, security related notification does not get buried within a load of marketing-pushes. 

2) Notification Timeout: We all know that most WordPress Systems are mostly living without attention for weeks at a time. I'd like notifications to have a possibility to define a timeout for when the notification becomes invalid and does not get shown: Example: A marketing push for Christmas does not make sense to be shown after the fact.

The intent here is to keep the notification list as clean and short as possible and to not clutter it with irrelevant information.

3) Comparisons to the Web Notifications API: One of the logical extensions for the WP Notify API is to attach Web or Push notifications. I would love the project to consider compatibility in terms of naming and usability of these existing APIs. 

See: https://developer.mozilla.org/en-US/docs/Web/API/Notification

**Comment by raaaahman** [comment]
Thanks @Luehrsen , those are great suggestions. I believe there is enough to talk about each of them to justify opening three distinct issues. Would you open those issues?

**Comment by raaaahman** [comment]
After tinkering with extending the Notification model (#61 and #63), I propose another way to think about it: we stick to only the required fields, and the `message` field is in fact either:

- A single paragraph to display
- A collection of data to display (images, links, titles, paragraphs)

![The Notification model is composed of a Message model that can either be a single or several fields](https://user-images.githubusercontent.com/16425382/128874811-6d89a096-d7dc-45c2-94ef-c0e133ca4333.png)

(White is what we have actually, and blue is what we would add)

This would mean that each element need to have its dedicated template part on the frontend side, and its dedicated model in the backend side. So, we would be able to add, delete, or modify such parts without needing to change the base Notification model.

The `message` could also be renamed `content` if it is more clear for everyone.

(I've added the `level` field, which could also be named `severity`, in the required fields. e.g.: 'info', 'warning', 'error', ...) 

EDIT: I am now wondering if the Status would need to included a timestamp from when the notification was read...

**Comment by amustaque97** [comment]
`Severity` looks more contextual and appealing to me. :smile: 
I'm confused here, `Message_Element` is part of the `message` property of `Base_Message`? 
If Yes, then cool :+1: 
If No, then how `Base_Message` is different from `Message_Element`?

> I am now wondering if the Status would need to include a timestamp from when the notification was read...

I think that's a good idea to store `timestamp` as well as `user` info who read notification first.

Status field will look something like this:
`{read: false, timestamp: null, user_id: null}`

Once that notification is read by someone data should look like this:
`{read: true, timestamp: 11-8-2021 4:10:30, user_id: 5}`

If it looks timestamp and user info is not required at the moment. We can introduce this in the coming future.

OR 

If we are storing multiple recipients then how we are going to make sure which user has read that notification. If we don't care whether someone has read the notification or not.. then we can ignore `timestamp` in the Status. Timestamp will be useful when we are maintaining some history.

**Comment by raaaahman** [comment]
You are right about the status, it should be related to each recipient. So either:

- Each notification is sent to a unique recipient:
  - CON: Each notification could be duplicated by the number of recipient it tries to reach
  - PRO: We can delete notifications as soon as its recipient request to do so
- OR the `status` is defined in a many-to-many relation between notification and recipients
  - PRO: No duplicated data
  - CON: We have to check that each recipient of a notification has read it/requested to erase it before we can delete it from the system

I would favor the first at the moment, as less complicated logic should bring less bug.

For your other question:

The `message` field of the notification could be either:
- a `Base_Message` which represent a single paragraph message
- a  `Message_Element_Collection` (could be named `Composite_Message` as well), which contains an undefined number of message elements. (Multi-paragraph, title, action link(s), etc.)

**Comment by amustaque97** [comment]
Thank You for the reply :)

First approach is fair enough 👍 

This is a follow-up regarding `message` field. If we have two types of messages ie. `Base_Message` and  `Composite_Message`. In my opinion it will be a bit complex to handle cases. Example suppose some plugin or theme sent an action notification as a `Base_Message` instead of `Composite_Message` or vice versa. Then there could be some possibility that things can break or it will not work as expected something like that.

**Comment by raaaahman** [comment]
The goal is that both `Base_Message` and `Composite_Message` implements the same interface `Message` interface, which extends `JsonSerializable` interface, so we can send them both to the frontend using the same method.

Now there are extra care to be taken indeed. Issues might occur if some developer try to add new fields to a `Base_Message` (images, title, etc.), which may result in an error. So either:

- Both `Base_Message` and `Composite_Message` instance are **immutable**, so if you want to add an extra field to an existing message, you'll have to create a new `Composite_Message` anyway/
- There is no `Base_Message` and we create only `Composite_Message`s, even if they are only composed of a single paragraph. (simpler, but we may need to enforce some kind of validation on `Composite_Message` body so it contains one mandatory paragraph)
